pasha/context.py

```diff
@@ -39,37 +39,83 @@ def __init__(self, num_workers):
+
+        if isinstance(extra_shape, int):
+            return (self.num_workers, extra_shape)
+        else:
+            return (self.num_workers,) + tuple(extra_shape)
+
+    @staticmethod
+    def _like_kwargs(prototype, dtype, order, shape):
+        if dtype is None:
+            dtype = prototype.dtype
+
+        order = order.upper()
+
+        if order in ('K', 'A'):
+            # Defining the branch in this order allows both cases to be
+            # handled.
+            order = 'F' if prototype.flags.f_contiguous else 'C'
+        elif order not in ('C', 'F'):
+            raise ValueError('invalid memory order')
+
+        if shape is None:
+            shape = prototype.shape
+
+        return dict(dtype=dtype, order=order, shape=shape)
+
+    def empty(self, shape, dtype=np.float64, order='C'):
         """Allocate an array shared with all workers.
 
         The implementation may decide how to back this memory, but it
         is required that all workers of this context may read AND write
-        to this memory. The base implementation allocates directly
-        on the heap.
+        to this memory. The array values are not guaranteed to be
+        initialized to any particular value.
 
         Args:
             shape (int or tuple of ints): Shape of the array.
-            dtype (data-type): Desired data-type for the array.
+            dtype (DTypeLike, optional): Data type of the array.
+            order ('C' or 'F', optional): Whether store multiple
+                dimensions in row-major (C-style, default) or
+                column-major (Fortran-style) order.
 
         Returns:
             (numpy.ndarray) Created array object.
         """
 
-        return np.zeros(shape, dtype=dtype)
+        raise NotImplementedError('Context.empty')
 
-    def array_like(self, other):
-        """Allocate an array with the same shape and dtype as another.
+    def empty_like(self, prototype, dtype=None, order='K', shape=None):
+        """Allocate a shared array with the same shape/dtype as another.
 
         Args:
-            other (ArrayLike): Other array.
+            prototype (ArrayLike): Array to copy properties from.
+            dtype (DTypeLike, optional): Override data type of the
+                resulting array.
+            order ('C', 'F', 'A' or 'K', optional): Override memory
+                order of the resulting array ('C' or 'F'), keep the
+                same layout ('K') or use Fortran-style if and only if
+                prototype is Fortran-style ('A').
+            shape (int or tuple of ints): Override shape of the
+                resulting array.
 
         Returns:
             (numpy.ndarray) Created array object.
         """
 
-        return self.array(other.shape, other.dtype)
+        return self.empty(**self._like_kwargs(prototype, dtype, order, shape))
 
-    def array_per_worker(self, shape, dtype=np.float64):
+    def empty_per_worker(self, shape, dtype=np.float64, order='C'):
```

Comment:
I presume it would normally make sense for each worker's section of the array to be contiguous in memory, which would mean adding the extra dimension at the *end* for Fortran-ordered data.

As this is only a convenience, maybe it's simpler to not offer configurable order for the `*_per_worker` methods, but just always return C order. It's not exactly hard for someone to construct an array with a suitable extra dimension if they do need to use it in Fortran order.

Reply:
🤦 
And there you see how often I use Fortran ordering.

Reply:
I've never used it and I don't think I've ever actually seen code that uses it... that's part of why I don't think it needs to be a priority.